codalab/bin/ws_server.py

```diff
@@ -1,75 +1,143 @@
 # Main entry point for CodaLab cl-ws-server.
 import argparse
 import asyncio
+from collections import defaultdict
 import logging
+import os
 import re
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 import websockets
+from dataclasses import dataclass
+import threading
 
-logger = logging.getLogger(__name__)
-logger.setLevel(logging.WARNING)
-logging.basicConfig(format='%(asctime)s %(message)s %(pathname)s %(lineno)d')
-
-worker_to_ws: Dict[str, Any] = {}
+from codalab.lib.codalab_manager import CodaLabManager
 
 
-async def rest_server_handler(websocket):
-    """Handles routes of the form: /main. This route is called by the rest-server
-    whenever a worker needs to be pinged (to ask it to check in). The body of the
-    message is the worker id to ping. This function sends a message to the worker
-    with that worker id through an appropriate websocket.
+@dataclass
+class WS:
+    """
+    Stores websocket object and whether or not the websocket is available.
+    TODO: give this a better, less confusing name.
     """
-    # Got a message from the rest server.
-    worker_id = await websocket.recv()
-    logger.warning(f"Got a message from the rest server, to ping worker: {worker_id}.")
 
-    try:
-        worker_ws = worker_to_ws[worker_id]
-        await worker_ws.send(worker_id)
-    except KeyError:
-        logger.error(f"Websocket not found for worker: {worker_id}")
+    _ws: Any = None
+    _is_available: bool = True
+    _lock: threading.Lock = threading.Lock()
+    _timeout: float = 86400
+    _last_use: Optional[float] = None
+
+    @property
+    def ws(self):
+        return self._ws
+
+    @property
+    def lock(self):
+        return self._lock
+
+    @property
+    def is_available(self):
+        return self._is_available
+
+    @is_available.setter
+    def is_available(self, value):
+        self._is_available = value
+
+    @property
+    def timeout(self):
+        return self._timeout
+
+    @property
+    def last_use(self):
+        return self._last_use
 
+    @last_use.setter
+    def last_use(self, value):
+        self._last_use = value
 
-async def worker_handler(websocket, worker_id):
-    """Handles routes of the form: /worker/{id}. This route is called when
-    a worker first connects to the ws-server, creating a connection that can
-    be used to ask the worker to check-in later.
+
+worker_to_ws: Dict[str, Dict[str, WS]] = defaultdict(
+    dict
+)  # Maps worker to list of its websockets (since each worker has a pool of connections)
+ACK = b'a'
+logger = logging.getLogger(__name__)
+manager = CodaLabManager()
+bundle_model = manager.model()
+worker_model = manager.worker_model()
+server_secret = os.environ["CODALAB_SERVER_SECRET"]
```

Comment:
We use a server secret (some cryptographically secure string) that would be an environment variable added to the deployment repo